### ISSUE TITLE
Structured output (json_schema) never terminates on the VLM engine — GrammarMatcher rejected token 0 every step until max_tokens

### ISSUE THREAD

**Issue by 7hdmm9pp2w-code** [jundot/omlx#3550]
<html><head></head><body><p><strong>
<h2>Summary</h2>
<p>Any request with <code>response_format: {"type": "json_schema", ..., "strict": true}</code> against a model served by <code>VLMBatchedEngine</code> loops until it hits <code>max_tokens</code>. The same request against a model served by <code>BatchedEngine</code> works correctly.</p>
<p>No image is required to reproduce — a plain text prompt to a VLM-backed model is enough, so this is not about image handling.</p>
<h2>Environment</h2>
<ul>
<li>oMLX 0.6.4, Homebrew install on macOS, Apple Silicon (M5 Pro, 128 GB)</li>
<li>xgrammar present via <code>brew reinstall omlx --with-grammar</code></li>
<li>Grammar works: <code>GrammarCompiler initialized for ...</code> appears for both engines</li>
</ul>
<h2>Reproduction</h2>
<pre><code class="language-bash">curl -s -m 120 http://localhost:8000/v1/chat/completions \
 -H 'Content-Type: application/json' -d '{
 "model":"Qwen3-VL-8B-Instruct-8bit",
 "messages":[{"role":"user","content":"2 EL Olivenöl"}],
 "response_format":{"type":"json_schema","json_schema":{"name":"z","strict":true,
  "schema":{"type":"object","properties":{
    "quantity":{"type":"number"},"unit":{"type":"string"},"food":{"type":"string"}},
  "required":["quantity","unit","food"],"additionalProperties":false}}}}'
</code></pre>
<h2>Actual</h2>
<p>The log fills with one warning per decode step, roughly every 20 ms, always token 0:</p>
<pre><code>omlx.api.grammar - WARNING - GrammarMatcher rejected token 0
omlx.api.grammar - WARNING - GrammarMatcher rejected token 0
...
omlx.server - INFO - Chat completion: model=Qwen3-VL-8B-Instruct-8bit,
  4096 tokens in 88.37s (46.4 tok/s), prompt: 616,
  finish_reason=length, max_tokens=4096
</code></pre>
<p>Generation only stops at <code>max_tokens</code>. The response is unusable.</p>
<h2>Expected</h2>
<p>Valid schema-conforming JSON with <code>finish_reason=stop</code>, as produced by the text engine.</p>
<h2>Affected / unaffected</h2>

Model | Engine | Result
-- | -- | --
Qwen3-VL-8B-Instruct-8bit | VLMBatchedEngine | loops to max_tokens
Qwen3-VL-4B-Instruct-8bit | VLMBatchedEngine | loops to max_tokens
Qwen3.6-27B-MLX-8bit | VLMBatchedEngine | loops to max_tokens
Qwen3-4B-Instruct-2507-8bit | BatchedEngine | works, finish_reason=stop

<p>The split follows the engine, not the model family or the quantization.</p>
<h2>Hypothesis</h2>
<p>In <code>omlx/api/grammar.py</code>, <code>_apply_mask(mx_bitmask, logits, self._vocab_size)</code> masks logits against a vocabulary size taken from the text tokenizer. VLM checkpoints extend the vocabulary with vision/image tokens, so on the VLM path the mask and the logits are misaligned. The sampled token is therefore always rejected:</p>
<pre><code class="language-python">if not self._matcher.accept_token(token_id):
    logger.warning("GrammarMatcher rejected token %d", token_id)
</code></pre>
<p><code>accept_token</code> failing leaves the matcher state unchanged, so the next step repeats identically and the loop never advances. That the rejected id is consistently 0 suggests the masked logits collapse to index 0.</p>
<h2>Impact</h2>
<p>Any OpenAI-compatible client that uses the structured-output helper (<code>client.chat.completions.parse</code>) is blocked from using VLM-served models. In our case this is Mealie's recipe import, whose image path always goes to a VLM provider — image import is unusable while the text path works fine.</p>
<h2>Workaround</h2>
<p>Route structured-output requests to a text-only model. There appears to be no way to disable grammar enforcement per model or per request short of omitting <code>response_format</code>, which clients like Mealie do not offer. Setting <code>sampling.max_tokens</code> to a low value at least bounds the hang.</p></body></html>
